src/main/kotlin/app/RepoHasher.kt

```diff
@@ -3,83 +3,196 @@
+        if (localRepo.remoteOrigin.isNotBlank()) {
+            repoRehash += localRepo.remoteOrigin
+        } else {
+            repoRehash += localRepo.path + localRepo.userName
+        }
+
+        repo.rehash = DigestUtils.sha256Hex(repoRehash)
+    }
+
+    private fun getRepoConfig() {
+        gitRepo ?: return
+        val config = gitRepo.getConfig()
+        localRepo.author = Author(
+                name = config.getString("user", null, "name") ?: "",
+                email = config.getString("user", null, "email") ?: "")
+        localRepo.remoteOrigin = config.getString("remote", "origin", "url")
+                ?: ""
+        localRepo.userName = try { System.getProperty("user.name") }
+                             catch (e: Exception) { "" }
+    }
+
+    private fun isKnownRepo(): Boolean {
+        if (Configurator.getRepos().find { it.rehash == repo.rehash } != null) {
+            return true
+        }
+        return false
+    }
+
+    private fun rehashNewCommits() {
+        addedCommits = mutableListOf()
+        removedCommits = repo.commits.toMutableList()
+        val lastServerCommit: Commit = repo.commits.last()
+        var isLastServerCommitChecked: Boolean = false
+
+        val commitsObservable = getObservableRevCommits()
+        commitsObservable.map(commitMapper).filter(emailFilter)
+                .filter { !isLastServerCommitChecked }
+                .blockingSubscribe({  // OnNext
+            Logger.info("Commit: ${it.rehash}")
+            if (it == lastServerCommit) {
+                isLastServerCommitChecked = true
+            }
+            if (removedCommits.contains(it)) {
+                removedCommits.remove(it)
+            } else {
+                addedCommits.add(it)
+            }
+        }, { t ->  // OnError
+            Logger.error("Error while hashing: $t")
+        })
+    }
+
+    private fun rehashAllCommits() {
+        addedCommits = mutableListOf()
+
+        val commitsObservable = getObservableRevCommits()
+        commitsObservable.map(commitMapper).filter(emailFilter)
+                .blockingSubscribe({  // OnNext
+            Logger.info("Commit: ${it.rehash}")
+            addedCommits.add(it)
+        }, { t ->  // OnError
+            Logger.error("Error while hashing: $t")
+        })
+    }
+
+    private fun getRepoFromServer() {
+        repo = SourcererApi.getRepo(repo.rehash)
+    }
+
+    private fun sendRepoToServer() {
+        SourcererApi.postRepo(repo)
+    }
+
+    private fun sendAddedCommits() {
+        if (addedCommits.isNotEmpty()) {
+            SourcererApi.postCommits(addedCommits)
+        }
+    }
+
+    private fun sendRemovedCommits() {
+        if (removedCommits.isNotEmpty()) {
+            SourcererApi.deleteCommits(removedCommits)
+        }
+    }
 
-        if (git != null) {
-            val repository = git.repository
+    private fun getObservableRevCommits(): Observable<RevCommit> =
```

Comment:
`just` will create observable from existing list, while using `create` we could start hashing commit right after reading it from repository.